drake/solvers/Optimization.h

```diff
@@ -284,9 +294,42 @@ namespace Drake {
       virtual MathematicalProgram* addNonlinearConstraint() { return new MathematicalProgram; };
       virtual MathematicalProgram* addLinearInequalityConstraint() { return new MathematicalProgram; };
      */
-      virtual MathematicalProgram* addLinearEqualityConstraint() { return new MathematicalProgram; };
+      virtual MathematicalProgram* addLinearEqualityConstraint() {
+        return require_superclass_of<LeastSquares>();
+	  };
 
       virtual bool solve(OptimizationProblem& prog) { throw std::runtime_error("not implemented yet"); }
+
+    private:
+      /**
+       * Returns this if it's a superclass of Derived. Otherwise constructs a Derived from
+       * provided arguments and deletes this.
+       * Derived must be in the same inheritance tree as this.
+       *
+       * @tparam Derived Class to compare this to
+       * @param derived_ctor_args arguments to construct a Derived, if necessary to do so
+       */
+      template <typename Derived, typename... DerivedCtorArg>
+      MathematicalProgram* require_superclass_of(DerivedCtorArg&&... derived_ctor_arg) {
+        //If *this is a superclass of Derived, we won't be able to cast *this into a Derived.
+        //An exception is that *this may be of type Derived, so we have check for this case
+        //We use this logic since C++ can't directly check if a static type derives from a dynamic type
+
+        bool this_is_superclass_of_derived =
+                (typeid(Derived) == typeid(*this)) //check if *this is a Derived
+                || (dynamic_cast<Derived*>(this) == nullptr); //check if we cannot cast this to Derived
+
+        if (this_is_superclass_of_derived)
+          return this;
+        else
+        {
+          //delete quad is safe IF quad is no longer used, as classes can delete themselves.
+          //however, perhaps this behaviour might be surprising to users?
+          //it's clean and concise, but potentially contains a serious pitfall
+          delete this;
```

Comment:
i think the unique_ptr reset would go here instead.  i think that labeling the program as unique_ptr (instead of the raw pointer) makes this deletion ok.